common/djangoapps/third_party_auth/api/views.py

```diff
@@ -89,3 +98,123 @@ def get(self, request, username):
+
+    **Query Parameters**
+
+        * remote_ids: Optional. List of comma separated remote (third party) user IDs to filter the result set.
+          e.g. ?remote_ids=8721384623
+
+        * usernames: Optional. List of comma separated edX usernames to filter the result set.
+          e.g. ?usernames=bob123,jane456
+
+        * page, page_size: Optional. Used for paging the result set, especially when getting
+          an unfiltered list.
+
+    **Response Values**
+
+        If the request for information about the user is successful, an HTTP 200 "OK" response
+        is returned.
+
+        The HTTP 200 response has the following values:
+
+        * count: The number of mappings for the backend.
+
+        * next: The URI to the next page of the mappings.
+
+        * previous: The URI to the previous page of the mappings.
+
+        * num_pages: The number of pages listing the mappings.
+
+        * results:  A list of mappings returned. Each collection in the list
+          contains these fields.
+
+            * username: The edx username
+
+            * remote_id: The Id from third party auth provider
+    """
+    authentication_classes = (
+        OAuth2Authentication,
+    )
+
+    paginate_by = 50
+    paginate_by_param = 'page_size'
+    pagination_serializer_class = PaginationSerializer
+    serializer_class = serializers.UserMappingSerializer
+    provider = None
+
+    def get_queryset(self):
+        self.provider = Registry.get(self.kwargs.get('provider_id'))
+        if not self.provider:
+            raise Http404
+
+        # permission checking. We allow both API_KEY access and OAuth2 client credential access
+        if not (
+                ApiKeyHeaderPermission().has_permission(self.request, self) or
+                ThirdPartyAuthApiPermission().has_permission(self.request, self)
+        ):
+            raise exceptions.PermissionDenied()
+
+        usernames_str = self.request.QUERY_PARAMS.get('usernames', None)
+        remote_ids_str = self.request.QUERY_PARAMS.get('remote_ids', None)
+
+        query_set = UserSocialAuth.objects.select_related('user').filter(provider=self.provider.backend_name)
+
+        # build our query filters
+        query = Q()
+        if usernames_str:
+            usernames = usernames_str.split(',')
+            query = query | Q(user__username__in=usernames)
+        if remote_ids_str:
+            remote_ids = [self.provider.get_social_auth_uid(remote_id) for remote_id in remote_ids_str.split(',')]
+            query = query | Q(uid__in=remote_ids)
+        if not usernames_str and not remote_ids_str:
+            # if no filter on usernames and remote ids, we need to add a prefix filter on uid so that in case of
+            # multi-IdP backend, we only retrieve the ones that are for current IdP
+            uid = self.provider.get_social_auth_uid('uid')
+            if uid is not 'uid':
+                query = query | Q(uid__startswith=uid[:-3])
```

Comment:
This filter should always be applied, not just when no `username_str` or `remote_ids_str` is present (and use `&` not `|`).

I have two SAML providers, `testshiba` and `testshibB`, and the `myself` user has connected his account to both of those. His entries in the UserSocialAuths table like this:
```
mysql> SELECT username, provider, uid FROM social_auth_usersocialauth, auth_user WHERE auth_user.id=user_id and user_id=16;
+-----------+----------+------------------+
| username  | provider | uid              |
+-----------+----------+------------------+
| myselfEdX | tpa-saml | testshibB:myself |
| myselfEdX | tpa-saml | testshiba:myself |
+-----------+----------+------------------+
```

When I run the command `curl --header "Authorization: Bearer 675..." http://localhost:8000/api/third_party_auth/v0/providers/saml-testshiba/users?usernames=myselfEdX` 

I get this traceback:

```
Traceback (most recent call last):
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 109, in get_response
    response = callback(request, *callback_args, **callback_kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/rest_framework/compat.py", line 137, in view
    return self.dispatch(request, *args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/views/decorators/csrf.py", line 77, in wrapped_view
    return view_func(*args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/rest_framework/views.py", line 400, in dispatch
    response = self.handle_exception(exc)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/rest_framework/views.py", line 397, in dispatch
    response = handler(request, *args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/rest_framework/generics.py", line 425, in get
    return self.list(request, *args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/rest_framework/mixins.py", line 98, in list
    return Response(serializer.data)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/rest_framework/serializers.py", line 575, in data
    self._data = self.to_native(obj)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/rest_framework/serializers.py", line 351, in to_native
    value = field.field_to_native(obj, field_name)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/rest_framework/serializers.py", line 410, in field_to_native
    return [self.to_native(item) for item in value.all()]
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/rest_framework/serializers.py", line 351, in to_native
    value = field.field_to_native(obj, field_name)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/rest_framework/fields.py", line 1035, in field_to_native
    value = getattr(self.parent, self.method_name)(obj)
  File "/edx/app/edxapp/edx-platform/common/djangoapps/third_party_auth/api/serializers.py", line 23, in get_remote_id
    return self.provider.get_remote_id_from_social_auth(social_user)
  File "/edx/app/edxapp/edx-platform/common/djangoapps/third_party_auth/models.py", line 319, in get_remote_id_from_social_auth
    assert self.match_social_auth(social_auth)
AssertionError
```

The reason is that my URL specifies the `testshiba` provider only, but the result set contains two rows, and the serializer asks `testshiba` to serialize both. But `testshiba` asserts when asked to serialize the UID `testshibB:myself`